### PR TITLE
Cauldrons and Bucket / Bottle interaction (and levers/buttons....)

### PR DESCRIPTION
**Pull request by Lindsay-Needs-Sleep**
First time messing around in a mod, so feedback welcome! 

This attempts to use the selected item on the hit block on `place`.  (Only attempts if the block is right beside/above/below to not interfere with actually placing blocks)

This enables:  
- turtles to fill/empty their buckets/bottles at cauldrons. (My main purpose)
- It also has the side effect of enabling turtles to interact with adjacent lever/buttons.
- You can also collect honey from bee hives with bottles, and shear honeycombs with shears (though they drop as items and don't directly go into the turtle's inventory)

Consequently changed behaviors:  
- If a chest is right beside a turtle, and you `place` a water bucket on it, nothing happens. (Previously it would waterlog the chest.)  But, you can still waterlog chests by calling `place` when the chest is not immediately beside the turtle.

Feedback:  
- Maybe there is something already existing to replace `areAdjacent`?
  - Or a better name? o.o;;
- Can anyone think of any other unintended side effects?

---
Tests

`./gradlew check` No new tests fail.  (These 2 failed before I made any changes, so I don't think it was me)

```
> Task :core:test

ComputerTestDelegate > get() > The os library > os.date > produces output consistent with PUC Lua > dan200.computercraft.core.ComputerTestDelegate.get()[73][1][1][26] FAILED
    org.opentest4j.AssertionFailedError at ComputerTestDelegate.java:420

ComputerTestDelegate > get() > The os library > os.date > produces output consistent with PUC Lua > dan200.computercraft.core.ComputerTestDelegate.get()[73][1][1][30] FAILED
    org.opentest4j.AssertionFailedError at ComputerTestDelegate.java:420
```

### CODE REVIEW

**Comment by SquidDev** [comment]
Thank you very much for working on this!

I should probably give some background on the turtle placing code, as the intention and implementation are pretty gnarly:

There's always been a bit of a discussion/debate about how much turtles should be able to interact with the world, and whether they should be able to right click arbitrary items/blocks (i.e. #125). My stance has generally been pretty conservative, in that turtles should be able to place blocks, and right click items into/against blocks (such as the cauldron and bee hive interactions), but shouldn't be able to use items and blocks arbitrarily (such as pressing buttons).

However, Minecraft's interface doesn't really allow this at all:
 - `Item.useOn` is used for placing blocks and items, that's fine. Except:
 - `Item.use` is used to place bucket of water and lillypad (which we want). However, it's also used to eat food and cast fishing rods (which we don't)[^1]. 
 - `Block.use` is used to place water into cauldrons (which we want), but again is also used to flip levers (which we don't).

This gets even messier as which method implements which interaction changes over time. Cauldrons _used_ to work until 1.18 (or 1.17, I can't remember!), as the interaction was moved from `Item.useOn` to `Block.use`.

---

The TLDR here is that turtles not being able to right click blocks _is_ intended, but also maybe the reasoning for that is silly and just causing more problems than it solves.

I'm afraid I'm going to sit on this for a few days and have a think before reviewing this any further. Thoughts from other people very much welcome!

Very sorry for the delay, especially on your first modding PR!

[^1]: We used to limit `Item.use` to a couple of specific items. I relaxed that in 8a7156785deaf49af054fbfa06671f314ff9857b, because the hard coding upset me, but it does now have the interesting behaviour that you can throw an ender pearl from a turtle. Albeit only once - we don't handle cooldowns properly.

**Comment by Lindsay-Needs-Sleep** [comment]
Np!  Thank you for the beautifully formatted (and informative comment!)

I must say I share the same sentiment (bucket/beehives: yes, levers/buttons/etc: no).

Although it would be fairly amusing seeing a turtle going around pressing buttons and flipping levers, this is really not necessary since we already have the redstone api.  Additionally, it's just kind of funny to call `place` to use a button (with the requirement that the turtle is holding an item like a bucket). xD

I was hoping it was just my lack of knowledge, and there would be better way to filter than `if cauldron/beehive` but it sounds like that is not the case.  
(Big thank you for the function breakdown!  It confirmed what I *thought* I was understanding.)

The other thing I just tested, this allows using any hoe in the selected slot with `place` to till the ground.  This does have the draw back of using up durability, but allows you to skip the diamond requirement (not really a fan of that).  But I suppose this is also fairly analogous to sheering sheep with `place` (which currently works).

The linked issue is interesting.  If there was a chance of `use` coming in, I must say I kind of agree with making any kind of bucket use fall under a `use` (vs. `place`).  Limiting it to certain items/interaction makes sense.  The dislike for hardcoding the items, makes even more sense though!  Opening it up to let it `use` whatever is an interesting thing.  I suppose it would give players a lot more freedom with their turtles, but it does remove some of the limitations, which are admittedly a fun part. :p  

This is certainly a fairly substantial consideration, so please take your time!

**Comment by MineRobber9000** [comment]
Adding my two cents where it wasn't asked for again, but you could probably use tags to decide what items can be used and on what blocks (`computercraft:turtle_use_items` and `computercraft:turtle_use_blocks` or similar naming). That way, if someone wants their turtles to be able to press buttons/flip levers/whatever, they can just add `#minecraft:buttons` and `minecraft:lever` (or whatever the button tag and lever item ID are resp.) to `computercraft:turtle_use_blocks` with a datapack, and the base mod can just handle cauldrons and beehives and whatever else we want turtles to be able to interact with without changes.

**Comment by SquidDev** [comment]
Sorry for the slow response here - tried to take the last week off and work on some other projects.

I had considered tags and then discounted them as they're still somewhat hard-coding a list. However, @MineRobber9000's point about them being user-customisable is something I hadn't considered before, and I think probably a good solution to this.

I've switched the usage of `Item.use` to check a tag in 2e5cd29e12db67c925c4ce5bd114c62d1bbae438. @Lindsay-Needs-Sleep, would you be willing to merge that into this branch and then do the same for `Block.use`? I imagine we'd just want cauldrons and hives in there right now - not sure what else is helpful!

**Comment by Lindsay-Needs-Sleep** [comment]
Sorry I seemed to have missed your comment somehow.   D:
My dev environment got nuked for this, I'm not sure when I'll have time again. >.<

But the tags sound great!
